docs/RFCS/certificate_rotation.md

```diff
@@ -0,0 +1,293 @@
+
+CA certificates can be added to the certificate pool with little extra validation.
+
+Node certificates must be checked for validity. Specifically:
+* we must have a valid certificate/private key pair
+* the certificate must currently be valid (`Not Before < time.Now() < Not After`)
+* the certificate must be signed by one of the CA certificates on this node
+
+The last condition is an attempt to avoid loading a certificate that may not be verifiable
+by other nodes or clients. If we do not have the right CA, chances are someone else does no either.
+
+We may need to set a timer for certificates that have not reached their `Not Before` date, otherwise
+we would need to trigger a second refresh.
+
+## Triggering a reload
+
+Running nodes can be told to re-read the certificate directory by issuing a `SIGHUP` to the process.
+
+Since we cannot control when nodes may be restarted, it is important to keep the reload process
+identical to the initial load.
+
+# Online certificate rotation
+
+A good description of online key rotation in Go can be found in
+[Hitless TLS Certificate Rotation in Go](https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/)
+
+## Adding a new CA certificate
+
+All TLS connections are given a [tls.Config](https://golang.org/pkg/crypto/tls/#Config) object
+containing a [*x509.CertPool](https://golang.org/pkg/crypto/x509/#CertPool) to verify
+server certificates (`tls.Config.RootCAs`) and client certificates (`tls.Config.ClientCAs`).
+
+Adding a new certificate to the pool can be done as long as we have access to the `tls.Config` or
+the `x509.CertPool`. Neither is safe for concurrent use so we will need a small wrapper to allow
+safe modification of the pool while serving connections.
+
+## Rotating node certificate
+
+Node certificates are tricker than CA certificates as we must pick exactly one to present.
+
+The `tls.Config` object allows us to specify node certificates in two ways:
+* set the certificate in `tls.Config.Certificates []Certificate`
+* set a per-connection callback in `tls.Config.GetCertificate func(*ClientHelloInfo) (*Certificate, error)`
+
+Modifying either will change which certificate is presented at session initiation or renegotiation.
+The `Certificates` object is not safe for concurrent use, making `GetCertificate` preferable.
+
+# Additional interfaces
+
+## Command line
+
+We will need to modify all commands that use certs to make use of the new directory structure.
+
+We will also need:
+* modification to `cert create-ca` to use an existing key.
+* `cert list` to list all CA certs and node/client cert/key pairs.
+
+## Admin UI / debug pages
+
+We want at least barebones listing of all certs on a given node, including validity
+dates, certificate chain (corresponding CA for a node cert), and valid hosts.
+
+Soon-to-expire certificates (or chains) must be reported prominently on the admin UI and
+available through external metrics.
+
```

Comment:
I would suggest also an UI form where the user can upload new certificates, that would subsequently be pushed to all the nodes and copied to their certs sub-dir.